### DISCUSSION TITLE
More mapper support for 2600 (and others)

### DISCUSSION THREAD

**Discussion by PsychoFox11** [sanni/cartreader#938] [Ideas & Suggestions]
Thanks to @smesgr9000 we've got great support for newer 7800 games (and may other improvements). Really appreciated btw! However, since I'm more of a ROM analyzer than someone who writes/understands dumping scripts (but am great at QA'ing them!), I'd love to see support for some of the Atari 2600 aftermarket mappers used by AtariAge mainly, and likely others.

If I'm making a dream wishlist, ColecoVision also has some more recent aftermarkets that aren't dumpable with current equipment (I also have the AtariMax Maxflash, but OSCR is beating it in support and accuracy now though ColecoVision is pretty equal). I also have a 192kb Vectrex game (Vectorblade) I'd love to dump, but I'm not holding my breath on support for a single game on an obscure system (we're lucky we got Demons of Asteborg support on Genesis, but that's pretty popular).

I know that sakman who wrote most of these typically focuses on retail which is fair, so hopefully we can convince someone to give the aftermarkets some love here too. The 7800 update was unexpected and very welcome, allowed me to do a couple I couldn't before.

**Comment by smesgr9000** [comment]
what 2600 mappings you are talking about? Melody/Harmony stuff as far as I know can't be dumped completely because you can't access the ARM-code via the cartridge port. Afaik this is the most common homebrew 2600 layout at the moment.
There are some notes on AtariAge that Melody is capable to speed up production of cartridge but I'm unsure why. Maybe reading the ARM code is possible but not documented.

**Reply by PsychoFox11** [reply]
Hmm yeah I'm not really sure, I don't have an intimate knowledge, but was looking at some recent releases like Grizzards etc. That I'm not sure what they use (even if I get a dump that loads I'm not sure if it is right). If it can't work from the edge connector, well it seems more are going in that direction.
I have heard some ColecoVision retails can't be fully dumped from the edge connector so I imagine aftermarkets are doing similar perhaps.

**Reply by PsychoFox11** [reply]
It does make me wonder though - I thought the 2600+ supported most if not all of what AtariAge sells - is that not the case? From my understanding it is just a dumper connected to an emulator. I wonder if it would hold any clues. Or maybe some of these just don't work with it and I didn't know that.

**Reply by smesgr9000** [reply]
First of all what is the difference between Melody & Harmony? AFAIK Harmony is a full fledged SDcard to 2600 adapter. While Melody is a simplified version of Harmony which replace the SD card with an Flash chip but otherwise is identical. I don't have a Harmony cartridge so I can't verify.

The point is Melody/Harmony can emulate a lot of mapper functions. The Melody/Harmony specific mapper called "DPC+" can't 
be dumped. Other should be possible. The problem with the emulate stuff - in case of Melody either by trial and error or by open the cartridge and read the configuration file.
The only mapper Harmony can do cartreader can't is "0840 Econobanking" and "Arcadia Supercharger". While the first one shouldn't be challenging if I would have a cart the second one is afaik out of target of this project. Afaik it would be easier to dump the tapes with a tape drive.

**Reply by smesgr9000** [reply]
I did open up a cartridge of a melody cartridge. I'm not familiar with NXP LPC213x ARM chips but as I see it the cart can't be read (or reprogrammed) without open up the cartridge. To get to the bootloader of the ARM the P0.14 must be low on reset. There is a SOT-23-chip not populated which imho could make it possible to archive that. If someone want to pickup from here feel free to contact me.
From the outside the following ports can be read/write:
P0.0-P0.12
P0.16-P0.23

**Reply by PsychoFox11** [reply]
Just circling back here. I was mainly trying to dump Grizzards, though I've also run into an aftermarket cart or two where the dumps work find, but don't match between Maxflash and OSCR so I'm not sure which to go with. It seems similar to an issue I pointed out a while back that Skaman addressed, with bytes at the end of a bank being from a different bank sometimes. I quit having the issue with retails doing that, but I have at least one, AtariAge's '2005 minigame multicart' that runs fine from either dumper, but bytes at the end of what I assume are banks don't match up, preventing me from recording it cause I dunno which is right!

**Reply by smesgr9000** [reply]
to be brutally honest - I would go with the AtariMax. With my prebuild hw5 I sometimes have some issues. I'm not 100% sure why but have two suspects:
- clock generator does odd things and is not really needed for 2600.
- 5V rail is not stable enough for me. The VCC goes down to 4.5-4.7V which some chips may not like and may produce odd stuff. It improves a bit if a USB charger is used but not much.

**Reply by PsychoFox11** [reply]
I'm sure it varies by chips but Atari dumped fine for me a while back when vSelect was broken and stuck on 3.3v without me knowing it. It was Coleco and Intellivision that made me notice and report the issue. But yeah with aftermarket chips they could be pickier or just different. I haven't figured out how to dump Grizzards at all though. I've had about as many issues with AtariMax as I have OSCR as far as a couple bytes being off or a game not working. I was only able to dump AnimAction for Vectrex from OSCR (only cart with RAM), but I could only do some 7800 aftermarkets from AtariMax - so it's hard to be definitive without desoldering a chip, which I'd rather not do on these, and they probably have SMD chips anyway which I don't have stuff to do.
